### PR TITLE
Switch to C# content type in cloud scenarios and remove C#_LSP content type and language

### PR DESCRIPTION
**Pull request by dibarbet**
TODO

- [x] Verify basic c# lsp functionality
- [x] Rebase on https://github.com/dotnet/roslyn/pull/45305
- [x] Test razor w/ live share
       This works about as well as it does in 16.7
- [x] Test razor w/ LSP (need to modify content type of LSP server they're looking for) - pr https://github.com/dotnet/aspnetcore-tooling/pull/2291
- [x] Test TS
- [x] Test debugging
       Saw #46403 but appears to be existing issue.
- [ ] Submit live share build with changes

### CODE REVIEW

File: eng/targets/Settings.props [hunk 1]
```diff
@@ -8,7 +8,7 @@
     <PackageTags>Roslyn CodeAnalysis Compiler CSharp VB VisualBasic Parser Scanner Lexer Emit CodeGeneration Metadata IL Compilation Scripting Syntax Semantics</PackageTags>
     <ThirdPartyNoticesFilePath>$(MSBuildThisFileDirectory)..\..\src\NuGet\ThirdPartyNotices.rtf</ThirdPartyNoticesFilePath>
 
-    <VSSDKTargetPlatformRegRootSuffix>RoslynDev</VSSDKTargetPlatformRegRootSuffix>
+    <VSSDKTargetPlatformRegRootSuffix>ExpLiveshare</VSSDKTargetPlatformRegRootSuffix>
```
Comment (dibarbet) [inline eng/targets/Settings.props:11]: TODO - Remove this before final PR.

File: src/VisualStudio/LiveShare/Impl/Client/Razor/CSharpLspContainedLanguageProvider.cs [hunk 2]
```diff
@@ -37,7 +36,7 @@ internal class CSharpLspContainedLanguageProvider : IContainedLanguageProvider
         }
 
         public IContentType GetContentType(string filePath)
```
Comment (dibarbet) [inline src/VisualStudio/LiveShare/Impl/Client/Razor/CSharpLspContainedLanguageProvider.cs:38]: Once razor switches fully to their LSP based editor in liveshare scenarios, this razor support can be removed.

File: src/VisualStudio/LiveShare/Impl/CustomProtocol/RoslynMethods.cs [hunk 3]
```diff
@@ -8,6 +8,5 @@ internal static class RoslynMethods
     {
         public const string ProjectsName = "roslyn/projects";
```
Comment (dibarbet) [inline src/VisualStudio/LiveShare/Impl/CustomProtocol/RoslynMethods.cs:9]: this is still required for typescript.  Once they switch to their LSP editor (~16.8ish) this method can also be removed.

File: src/VisualStudio/Core/Def/Implementation/LanguageClient/LiveShareLanguageServerClient.cs [hunk 4]
```diff
@@ -15,11 +15,11 @@
 
 namespace Microsoft.VisualStudio.LanguageServices.Implementation.LanguageService
 {
-    // The C#_LSP and VB_LSP ILanguageClient should not activate on the host. When LiveShare mirrors the C#_LSP ILC to the guest, they will not copy the DisableUserExperience attribute,
-    // so guests will still use the C#_LSP ILC.
+    // The C#P and VB ILanguageClient should not activate on the host. When LiveShare mirrors the C# ILC to the guest, they will not copy the DisableUserExperience attribute,
```
Comment (davidwengier) [inline src/VisualStudio/Core/Def/Implementation/LanguageClient/LiveShareLanguageServerClient.cs:18]:
```suggestion
    // The C# and VB ILanguageClient should not activate on the host. When LiveShare mirrors the C# ILC to the guest, they will not copy the DisableUserExperience attribute,
```

File: src/VisualStudio/Core/Def/Implementation/LanguageService/AbstractLanguageService`2.VsCodeWindowManager.cs [hunk 5]
```diff
@@ -128,10 +129,12 @@ private void AddOrRemoveDropdown()
...
-                // Temporary solution until the editor provides a proper way to resolve the correct navbar.
-                // Tracked in https://github.com/dotnet/roslyn/issues/40989
-                var document = _languageService.EditorAdaptersFactoryService.GetDataBuffer(buffer)?.AsTextContainer().GetRelatedDocuments().FirstOrDefault();
-                if (document?.GetLanguageService<INavigationBarItemService>() == null)
+                var textBuffer = _languageService.EditorAdaptersFactoryService.GetDataBuffer(buffer);
+                var document = textBuffer.AsTextContainer().GetRelatedDocuments().FirstOrDefault();
+                // TODO - Remove the TS check once they move the liveshare navbar to LSP.  Then we can also switch to LSP
+                // for the local navbar implementation.
+                // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1163360
+                if (textBuffer != null && textBuffer.IsInCloudEnvironmentClientContext() && document.Project.Language != "TypeScript")
```
Comment (davidwengier) [inline src/VisualStudio/Core/Def/Implementation/LanguageService/AbstractLanguageService`2.VsCodeWindowManager.cs:137]: If `textBuffer` was null, line 133 would have already thrown

File: src/EditorFeatures/Core.Wpf/InlineRename/CommandHandlers/RenameCommandHandler_RenameHandler.cs [hunk 6]
```diff
@@ -33,6 +31,11 @@ public CommandState GetCommandState(RenameCommandArgs args)
 
         public bool ExecuteCommand(RenameCommandArgs args, CommandExecutionContext context)
         {
+            if (!CanRename(args))
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core.Wpf/InlineRename/CommandHandlers/RenameCommandHandler_RenameHandler.cs:34]: Why was this change needed?

File: src/EditorFeatures/Core.Wpf/InlineRename/CommandHandlers/RenameCommandHandler_RenameHandler.cs [hunk 7]
```diff
@@ -99,6 +102,18 @@ private void ExecuteRenameWorker(RenameCommandArgs args, CommandExecutionContext
             }
         }
 
+        private static bool CanRename(RenameCommandArgs args)
+        {
+            if (!args.SubjectBuffer.TryGetWorkspace(out var workspace) ||
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core.Wpf/InlineRename/CommandHandlers/RenameCommandHandler_RenameHandler.cs:107]: Negate the condition in the if and then you can remove the if entirely?

File: src/Features/Core/Portable/Diagnostics/DiagnosticAnalyzerService_IncrementalAnalyzer.cs [hunk 8]
```diff
@@ -12,7 +12,7 @@ namespace Microsoft.CodeAnalysis.Diagnostics
 {
     [ExportIncrementalAnalyzerProvider(
         highPriorityForActiveFile: true, name: WellKnownSolutionCrawlerAnalyzers.Diagnostic,
-        workspaceKinds: new string[] { WorkspaceKind.Host, WorkspaceKind.Interactive, WorkspaceKind.CloudEnvironmentClientWorkspace })]
+        workspaceKinds: new string[] { WorkspaceKind.Host, WorkspaceKind.Interactive })]
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/Diagnostics/DiagnosticAnalyzerService_IncrementalAnalyzer.cs:15]: This doesn't break diagnostics in some way?

File: src/VisualStudio/Core/Def/Implementation/LanguageService/AbstractLanguageService`2.VsCodeWindowManager.cs (hunk 5 above)
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Implementation/LanguageService/AbstractLanguageService`2.VsCodeWindowManager.cs:137]: We already have the document for a workspace, any reason to be doing the lookup a second time with the textBuffer.IsInCloudEnvironmentClientContext() or should we just ask the workspace service from the document we already have?

File: src/VisualStudio/LiveShare/Impl/Shims/TypeScriptHandlerShims.cs [hunk 9]
```diff
@@ -117,76 +104,6 @@ public async Task<SymbolInformation[]> HandleAsync(DocumentSymbolParams param, R
...
-            => _threadingContext = threadingContext;
-
-        public Task<TextEdit[]> HandleAsync(DocumentOnTypeFormattingParams request, RequestContext<Solution> requestContext, CancellationToken cancellationToken)
-            => base.HandleRequestAsync(request, requestContext?.ClientCapabilities?.ToObject<ClientCapabilities>(), null, cancellationToken);
-
-        protected override async Task<IList<TextChange>?> GetFormattingChangesAsync(IEditorFormattingService formattingService, Document document, char typedChar, int position, CancellationToken cancellationToken)
-        {
-            // TypeScript expects to be called on the UI thread to get the formatting options.
-            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
-            return await base.GetFormattingChangesAsync(formattingService, document, typedChar, position, cancellationToken).ConfigureAwait(false);
```
Comment (jasonmalinowski) [inline src/VisualStudio/LiveShare/Impl/Shims/TypeScriptHandlerShims.cs:179]: Is something else doing this translation now?

File: src/VisualStudio/LiveShare/Impl/Client/LocalForwarders/VBRemoteLineSeparatorService.cs [hunk 10]
```diff
@@ -1,25 +0,0 @@
...
-using Microsoft.CodeAnalysis.Host;
-using Microsoft.CodeAnalysis.Host.Mef;
-using Microsoft.CodeAnalysis.Editor;
-using System.Composition;
-using System;
-
-namespace Microsoft.VisualStudio.LanguageServices.LiveShare.Client.LocalForwarders
-{
-    [ExportLanguageServiceFactory(typeof(ILineSeparatorService), StringConstants.VBLspLanguageName), Shared]
-    internal class VBLspLineSeparatorServiceFactory : ILanguageServiceFactory
```
Comment (jasonmalinowski) [inline src/VisualStudio/LiveShare/Impl/Client/LocalForwarders/VBRemoteLineSeparatorService.cs:14]: So we'll just have no line separator support anymore?

File: src/EditorFeatures/Core.Wpf/InlineRename/CommandHandlers/RenameCommandHandler_RenameHandler.cs (hunk 6 above)
Comment (dibarbet) [reply src/EditorFeatures/Core.Wpf/InlineRename/CommandHandlers/RenameCommandHandler_RenameHandler.cs:34]: Otherwise, roslyn will try and execute the rename (fail) and return true, causing the LSP rename handler to not get invoked.

File: src/Features/Core/Portable/Diagnostics/DiagnosticAnalyzerService_IncrementalAnalyzer.cs (hunk 8 above)
Comment (dibarbet) [reply src/Features/Core/Portable/Diagnostics/DiagnosticAnalyzerService_IncrementalAnalyzer.cs:15]: No, this used to be there b/c the roslyn diagnostic service was running in the client (and calling the server to get diagnostics).  Now that we're using LSP diagnostics it doesn't need to run on the client.

File: src/VisualStudio/LiveShare/Impl/Shims/TypeScriptHandlerShims.cs (hunk 9 above)
Comment (dibarbet) [reply src/VisualStudio/LiveShare/Impl/Shims/TypeScriptHandlerShims.cs:179]: Yes, this is handled by typescript on the client already.

**Comment by JoeRobich** [comment]
I kicked off a new validation PR for you - https://devdiv.visualstudio.com/DevDiv/_git/VS/pullrequest/265606

**Comment by RikkiGibson** [comment]
@dibarbet we can merge this if you want.

**Comment by dibarbet** [comment]
@RikkiGibson Hold off on this one - this requires a triple insertion with liveshare + razor.  I'm waiting on doing that until our infra is good.

**Comment by dibarbet** [comment]
Merging in preparation of insertion tomorrow
